components/provisioner/internal/installation/resource_cleanup.go

```diff
@@ -46,86 +46,139 @@ type serviceCatalogClient struct {
 }
 
 func (s *serviceCatalogClient) PerformCleanup() error {
-	//1. Fetch all service cluster service brokers
+	// Fetch all ClusterServiceBrokers
+	//TODO: Add retries for listing resources in case api-server fails to respond
 	clusterServiceBrokers, err := s.ListClusterServiceBroker(metav1.ListOptions{})
 	if err != nil {
-		return errors.Wrapf(err, "while listing cluster service brokers")
+		return errors.Wrapf(err, "while listing ClusterServiceBrokers")
 	}
 
-	//2. Filter based on broker URL prefix
+	// Filter the list based on ClusterServiceBroker's URL prefix
+	brokersWithUrlPrefix := s.FilterCsbWithUrlPrefix(clusterServiceBrokers, BrokerUrlPrefix)
+
+	// Get all ClusterServiceClasses from filtered ClusterServiceBrokers
+	cscWithMatchingLabel, err := s.GetClusterServiceClassesForBrokers(brokersWithUrlPrefix)
+	if err != nil {
+		return errors.Wrapf(err, "while getting ClusterServiceClasses")
+	}
+
+	// Get ServiceInstances for each ClusterServiceClass
+	serviceInstances, err := s.GetServiceInstancesForClusterServiceClasses(cscWithMatchingLabel)
+	if err != nil {
+		return errors.Wrapf(err, "while getting ServiceInstances")
+	}
+
+	err = s.DeleteServiceInstances(serviceInstances)
+	if err != nil {
+		return errors.Wrapf(err, "while deleting ServiceInstances")
+	}
+
+	return nil
+}
+
+func (s *serviceCatalogClient) ListClusterServiceBroker(options metav1.ListOptions) (*v1beta1.ClusterServiceBrokerList, error) {
```

Comment:
First of all the List* functions doesn't need to be public. You could also consider getting rid of those  and call Service Catalog client directly in FilterCsbWithUrlPrefix, GetClusterServiceClassesForBrokers, GetServiceInstancesForClusterServiceClasses, GetServiceInstancesForClusterServiceClasses.